segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/cache/CacheStorage.java

```diff
@@ -0,0 +1,122 @@
+     */
+    int insert(@NonNull BufferView data);
+
+    /**
+     * Replaces the entry at the given address with a new contents.
+     *
+     * @param address The address to replace. May be obtained via {@link #insert} or {@link #replace}. If there is no
+     *                entry with this address, this method will act like {@link #insert}.
+     * @param data    A {@link BufferView} representing the new contents of the entry. May be empty. Must have
+     *                {@link BufferView#getLength()} less than or equal to {@link #getMaxEntryLength()}.
+     * @return An integer representing the address where this data now resides. Depending on the implementation, this may
+     * or may not be equal to `address`. This can be used to invoke {@link #append}, {@link #get} or {@link #delete}.
+     * @throws CacheFullException If there is no more capacity in the {@link CacheStorage} to accommodate this entry.
+     * @throws CacheException     If another unexpected exception occurred.
+     */
+    int replace(int address, @NonNull BufferView data);
+
+    /**
+     * Gets the number of bytes that can be appended to an entry that has the given length.
+     *
+     * @param currentLength The current length of the entry.
+     * @return An non-negative integer representing the number of bytes that can be appended.
+     */
+    int getAppendableLength(int currentLength);
+
+    /**
+     * Performs an atomic compare-and-appends of data at the end of an entry. This operation will only take effect if
+     * the current length of the entry matches the argument `expectedLength` and both the entry's length and contents
+     * will be updated atomically.
+     *
+     * @param address        An integer representing the address to append to.
+     * @param expectedLength The currently known length of this entry.
+     * @param data           A {@link BufferView} representing the data to append.
+     * @return An integer representing the number of bytes that were appended. If 0, then the entry at the given address
+     * is already at its max capacity (its length equals {@link #getMaxEntryLength()}.
+     * @throws IncorrectCacheEntryLengthException If `expectedLength` does not match the current length of the entry.
+     * @throws IllegalArgumentException If address does not point to a valid entry or the given {@link BufferView} is
+     * too long to fit into the given entry (use {@link #getAppendableLength} to determine maximum length).
+     */
+    int append(int address, int expectedLength, @NonNull BufferView data);
+
+    /**
+     * Deletes the entry at the given address. If this address does not currently map to an entry, this operation will have
+     * no effect.
+     *
+     * @param address An integer representing the address to delete.
+     */
+    void delete(int address);
+
+    /**
+     * Retrieves the contents of an entry with the given address.
```

Comment:
I'm not clear on the expected behavior for this call, and specifically, my question is the following. Say that the address passed corresponds to an old view of the entry, and it maps to a different position in the last block or even to a different block altogether. What is `get` expected to return? A stale view of the entry corresponding to the passed address? The current view of the entry? An error?

Reply:
The address always points to a block and never to an offset within a block. No two entries may share a block.

Once inserted, an entry will never change its address. `append` is the only thing modifying an entry, but its purpose is to fill up the last block, but not to add more blocks to the entry. So there is no such thing as an "old view of the entry".

If the entry has been deleted, the address will either point to a "free" block (which will result in an error), or to some other entry, in which case that entry's contents will be returned. You may say that this is incorrect, but this is a low-level storage (just like disk blocks), and it's up to the upstream code (`ReadIndex`) to manage the entries and make sure that when `delete` is invoked it forgets about that address.

Reply:
Yes, if you still have an old address around, you can go read some garbage data from another entry (if it had been written again).